### PR TITLE
Make Array Field serialization consistent.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

#### Motivation

- `writeBinary(Array)` was inconsistent with `FieldVisitorWriteBinary`
- Looks like it was not used anywhere except `getBinaryValue`, which was required only for serialization of nested structures (Array, Map, Object, Tuple). So, the issue could have been possible only for Array inside another nested type, which is rare.

#### What is changed
- `writeBinary` is renamed to `writeBinaryArray`, which is consistent with  `FieldVisitorWriteBinary` now
- Added functions `writeFieldBinary`, `readFieldBinary` for generic fields.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 5728c3bb509c9875d2246cb7e9d243ea18e3a42a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70803/5728c3bb509c9875d2246cb7e9d243ea18e3a42a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/70607/286610f3f3dc5e6dee81b3ed4ef03134e1f55c8b/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by alexey-milovidov — APPROVED** [review]

**Review by Algunenano — APPROVED** [review]
The heavily templated code we have in the shape of ReadHelpers / WriteHelpers we have on top of this code makes me scary. The fact that you changed the name from `writeBinary` to `writeBinaryArray` and nothing required changes makes me confident the code was unused, so nothing can break by fixing this non-impactful bug. 

LGTM

File: src/Core/Field.h [hunk 1]
```diff
@@ -942,14 +942,15 @@ inline Field & Field::operator=(String && str)
 class ReadBuffer;
 class WriteBuffer;
 
-/// It is assumed that all elements of the array have the same type.
-void readBinary(Array & x, ReadBuffer & buf);
+/// Binary serialization of generic field.
+void writeFieldBinary(const Field & x, WriteBuffer & buf);
+Field readFieldBinary(ReadBuffer & buf);
+
+void readBinaryArray(Array & x, ReadBuffer & buf);
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Core/Field.h:949]: I think so.
My reasoning was to show that the compiler did not use it, indeed.
Maybe we can rename others to make it consistent.

[the remark this answers is not in the document]
Comment (azat) [reply src/Core/Field.h:949]: Should this helpers even be exposed at all?

[the remark this answers is not in the document]
Comment (azat) [reply src/Core/Field.h:949]: AFAICS the following helpers were not used at all:
- `writeBinaryArray`
- `writeBinary(Tuple)`
- `writeBinary(Object)`

The only function that is used now is `writeBinary(Map)` and is required only for `additional_table_filters`, only for an old protocol that uses binary settings serialization, and at that time `Map` wasn't supported at all likely.

So, maybe it is better to remove this helpers completely, or write some tests to make sure that they are correct?

Or maybe I missing something?

**Pull request merged**